secretsharing/ss.go

```diff
@@ -0,0 +1,183 @@
+
+// Recover returns the secret provided more than t shares are given. Returns an
+// error if the number of shares is not above the threshold t.
+func (s SecretSharing) Recover(shares []Share) (group.Scalar, error) {
+	if l := len(shares); l <= int(s.t) {
+		return nil, errThreshold(s.t, uint(l))
+	}
+
+	x := make([]group.Scalar, s.t+1)
+	px := make([]group.Scalar, s.t+1)
+	for i := range shares[:s.t+1] {
+		x[i] = shares[i].ID
+		px[i] = shares[i].Value
+	}
+
+	l := polynomial.NewLagrangePolynomial(x, px)
+	zero := shares[0].ID.Group().NewScalar()
+
+	return l.Evaluate(zero), nil
+}
+
+type SharesCommitment = []group.Element
+
+// VerifiableSecretSharing provides a (t,n) Feldman's verifiable secret sharing.
+// It allows splitting a secret into n shares, such that the secret can be only
+// recovered from any subset of t+1 shares.
+// It's verifiable as it allows checking whether a share is part of a secret
+// committed during sharding.
+type VerifiableSecretSharing struct{ s SecretSharing }
+
+// NewFeldmanSecretSharing implements a (t,n) Feldman's verifiable secret
+// sharing with threshold t.
+func NewFeldmanSecretSharing(t uint) (v VerifiableSecretSharing) { v.s.t = t; return }
+
+// Shard splits the secret into n shares, and also returns a commitment to both
+// the secret and the shares. The ShareCommitment must be sent to each party
+// so each party can verify its share is correct. Sharding a secret more
+// than once produces ShareCommitments with the same first entry.
+func (v VerifiableSecretSharing) Shard(rnd io.Reader, secret group.Scalar, n uint) ([]Share, SharesCommitment, error) {
+	splitter := NewSplitter(rnd, v.s.t, secret)
+	shares, err := splitter.multipleShard(n)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	g := secret.Group()
+	shareComs := make(SharesCommitment, splitter.poly.Degree()+1)
+	for i := range shareComs {
+		shareComs[i] = g.NewElement().MulGen(splitter.poly.Coefficient(uint(i)))
+	}
+
+	return shares, shareComs, nil
+}
+
+// Verify returns true if a share was produced by sharding a secret. It uses the
+// share commitments generated by the Shard function.
+func (v VerifiableSecretSharing) Verify(s Share, c SharesCommitment) bool {
+	if len(c) != int(v.s.t+1) {
+		return false
+	}
+
+	g := s.ID.Group()
+	lc := len(c) - 1
+	sum := g.NewElement().Set(c[lc])
+	for i := lc - 1; i >= 0; i-- {
+		sum.Mul(sum, s.ID)
+		sum.Add(sum, c[i])
+	}
+	polI := g.NewElement().MulGen(s.Value)
+	return polI.IsEqual(sum)
+}
+
+// Recover returns the secret provided more than t shares are given. Returns an
+// error if the number of shares is not above the threshold t.
+func (v VerifiableSecretSharing) Recover(shares []Share) (group.Scalar, error) {
+	return v.s.Recover(shares)
+}
+
+type Splitter struct {
+	g    group.Group
+	t    uint
+	poly polynomial.Polynomial
+}
+
+// NewSplitter returns a Splitter that can shard a secret with threshold t.
+func NewSplitter(rnd io.Reader, t uint, secret group.Scalar) (sp Splitter) {
```

Comment:
This interface isn't used in the tests -- what is the use case?

Reply:
removed